torch/csrc/jit/codegen/cuda/runtime/block_welford_outer.cu

```diff
@@ -0,0 +1,265 @@
+        self_var = inp_var[self_offset];
+      } else {
+        // Push second half
+        auto push_offset = index_in_chunk + chunk_size;
+        auto self_offset = index_in_chunk;
+        pushed_avg = inp_avg[push_offset];
+        pushed_var = inp_var[push_offset];
+        self_avg = inp_avg[self_offset];
+        self_var = inp_var[self_offset];
+      }
+      auto peer_avg = __shfl_xor_sync(0xffffffff, pushed_avg, lane_mask);
+      auto peer_var = __shfl_xor_sync(0xffffffff, pushed_var, lane_mask);
+
+      auto delta = peer_avg - self_avg;
+      self_avg += delta * b_N_div_ab_N;
+      self_var += peer_var + delta * delta * ((DataType)(inp_N)) * b_N_div_ab_N;
+
+      inp_avg[index_in_chunk] = self_avg;
+      inp_var[index_in_chunk] = self_var;
+    }
+    inp_N = updated_N;
+  }
+
+  // At this point, chunk_size is reduced to NumVals/wdimy as
+  // mentioned above. Each thread has warp-reduced chunk_size values
+  // in array inp. This chunk_size_post_reduction should be equal to
+  // chunk_size at this point.
+  constexpr int chunk_size_post_reduction = NumVals / wdimy;
+
+  // More specifically, the warp_tidy of each thread defines
+  // the chunk IDs held by the thread as follows:
+  //
+  // [warp_tidy * chunk_size_post_reduction, warp_tidy *
+  // chunk_size_post_reduction + chunk_size_post_reduction]
+  //
+  // Each thread uploads the chunk_size_post_reduction values one by
+  // one. Each chunk is spread by BDIMX * BDIMY values. The data
+  // layout of the shared memory is:
+  //
+  // [chunk_size, wid, warp_tidy, TIDx]
+  //
+  // The remaining reduction is done on the WID
+  // dimension. More specifically, we assign one warp per chunk (or
+  // a value of the group). The wdimy threads of the same threadId.x
+  // collectively reduce num_warps partial results, each of which is
+  // stored with stride 32. This means that there will be wdimy-way
+  // bank conflicts, so to avoid that, swizziling is also employed.
+#pragma unroll
+  for (int i = 0; i < chunk_size; ++i) {
+    // Accumulating smem offset from the innermost dimension
+    int smem_offset = 0;
+    // TIDx
+    smem_offset += threadIdx.x;
+    // Warp_TIDy with swizzle
+    smem_offset += ((warp_tidy + wid) % wdimy) * BDIMX;
+    // WID
+    smem_offset += wid * 32;
+    // chunk_size
+    smem_offset += i * BDIMX * BDIMY;
+    smem[smem_offset] = inp_avg[i];
+    smem[smem_var_offset + smem_offset] = inp_var[i];
+    // Upload N only when threadIdx.x == 0 && chunk_index == 0
+    if (threadIdx.x == 0 && i == 0 && warp_tidy == 0) {
+      reinterpret_cast<nvfuser_index_t*>(smem + smem_N_offset)[wid] = inp_N;
+    }
+  }
+
+  __syncthreads();
+
+  // The next step is to let each thread of a warp independently
+  // accumulate the partial results on the shared memory
+  // reduction. A single warp is used to accumulate of the partial
+  // results for a single chunk, so warp wid takes care of the wid-th
+  // chunk.
```

Comment:
Because now you have `[chunk_size, wid, warp_tidy, TIDx]` in shared memory, so you just want to iterate across `wid` but you might want to expand `wid` across `chunk_size` since they're independent?

Reply:
Would help to have consistent upper case extents and lower case iterators.

Reply:
Not sure what you meant by "expand `wid` across `chunk_size`", but maybe the below comment responses may help clarify.

Reply:
> Would help to have consistent upper case extents and lower case iterators.

Yeah, I just used upper cases only for template parameters, but that may make things easier to follow.